ipsec_offload_plugin/ipsec_offload/ipsec_offload.c

```diff
@@ -290,30 +292,73 @@ static void process_expire(ipsec_offload_config_queue_data_t *hdr)
 	}
 }
 
+static int fill_config_data(struct ipsec_offload_config_queue_data *data, char *buf)
+{
+   int proto, family;
+   char* rest = buf;
+   int soft_expire;
+   char* token;
+
+    while ((token = strtok_r(rest, ",", &rest))) {
+	char* key = strtok(token, ":");
+	char* value = strtok(NULL, ":");
+
+	if (!strcmp("ipsec-sa-spi", key))
+		sscanf(value, "%u", &data->spi);
+	else if (!strcmp(" soft-lifetime-expire", key)) {
+		sscanf(value, "%d", &soft_expire);
+		if (soft_expire)
+			data->cookie = SOFT_AGING_THRESHOLD_CROSSED;
+		else
+			data->cookie = HARD_AGING_THRESHOLD_CROSSED;
+	}
+	else if (!strcmp(" ipsec-sa-protocol", key)) {
+		sscanf(value, "%d", &proto);
+		data->proto = proto;
+	}
+	else if (!strcmp(" ipsec-sa-dest-address", key))
+		sscanf(value, "%s", data->dip);
+	else if (!strcmp(" address-family", key)) {
+		sscanf(value, "%d", &family);
+		if (family == 1)
+			data->family = IPv4;
+		else
+			data->family = IPv6;
+	}
+	else
+		return -1;
+    }
+    return 0;
+}
+
 void *audit_log_poll(void *arg) {
 	pthread_detach(pthread_self());
-	ipsec_offload_config_queue_data_u cfg_data;
+	struct ipsec_offload_config_queue_data cfg_data;
+	char config_data_buf[CONFIG_DATA_MAX];
+	int i;
 
 	while (1) {
 		if (*(bool *)arg == true) {
 			pthread_exit(NULL);
 		}
 
-		int ret = ipsec_fetch_audit_log(cfg_data.serialized_ipsec_offload_config_queue_data_t,
-									sizeof(ipsec_offload_config_queue_data_t));
+		int ret = ipsec_fetch_audit_log(config_data_buf, CONFIG_DATA_MAX);
 
 		if(ret != IPSEC_FAILURE)
 		{
-			cfg_data.config_queue_data.sip[IP_BUF] = '\0';
-			process_expire(&cfg_data.config_queue_data);
+			if (!fill_config_data(&cfg_data, config_data_buf)) {
+				DBG1(DBG_KNL, "audit_log data= spi=%u:proto=%d:dip=%s:family=%d:cookie=%d\n",
+				     cfg_data.spi, cfg_data.proto, cfg_data.dip, cfg_data.family, cfg_data.cookie);
+				cfg_data.sip[IP_BUF] = '\0';
+				process_expire(&cfg_data);
+			}
 		}
 		usleep(TEN_MS);
```

Comment:
shall we remove this sleep ? or make the sleep time same as backend manager. We should think to remove sleep and make it busy polling.

Reply:
removed it.